src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -753,6 +753,103 @@ fsm_stop_postgres(Keeper *keeper)
 }
 
 
+/*
+ * fsm_stop_postgres_for_maintenance is used when pg_autoctl enable maintenance
+ * has been used on the primary server, we do a couple CHECKPOINT before
+ * stopping Postgres to ensure a smooth transition.
+ */
+bool
+fsm_stop_postgres_for_maintenance(Keeper *keeper)
+{
+	LocalPostgresServer *postgres = &(keeper->postgres);
+	PostgresSetup *pgSetup = &(postgres->postgresSetup);
+	PGSQL *pgsql = &(postgres->sqlClient);
+
+	if (pg_setup_is_running(pgSetup))
+	{
+		/*
+		 * PostgreSQL shutdown sequence includes a CHECKPOINT, that is issued
+		 * by the checkpointer process one every query backend has stopped
+		 * already. During this final CHECKPOINT no work can be done, so it's
+		 * best to reduce the amount of work needed there. To reduce the
+		 * checkpointer shutdown activity, we perform a manual shutdown while
+		 * still having concurrent activity.
+		 *
+		 * The first checkpoint writes all the in-memory buffers, the second
+		 * checkpoint writes everything that was added during the first one.
+		 */
+		log_info("Preparing shutdown for maintenance: CHECKPOINT;");
+
+		/* we may fail to CHECKPOINT and then will retry at next FSM loop */
```

Comment:
Do we really want this? If checkpoint fails, maybe we simply want to quit without doing one.
I assume it's not common for CHECKPOINT to fail. So in the rare case that it does, might fail continuously if we keep retrying. Then we never reach maintenance state.

Reply:
CHECKPOINT may fail because File System Is Full, in which case a maintenance operation is needed on the primary. So yeah I think you're right, we should try to checkpoint twice but accept that it may fail, because the reason for the CHECKPOINT failure might also be the reason why an operator wants to put this node in maintenance...

Reply:
Done that way now (errors are only a warning and we continue forward with the stopping of Postgres).